File: app/src/main/java/ua/hospes/rtm/core/StopWatchService.java
package ua.hospes.rtm.core;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import android.os.PowerManager;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;

import ua.hospes.rtm.R;
import ua.hospes.rtm.ui.MainActivity;
import ua.hospes.rtm.utils.TimeUtils;

/**
 * @author Andrew Khloponin
 */
public class StopWatchService extends Service implements StopWatch.OnStopWatchListener {
    private static final int NOTIFICATION_ID = 1337;
    private static final String KEY_ACTION = "action";
    private static final String KEY_START = "start";
    private static final String KEY_STOP = "stop";
    private final IBinder binder = new StopWatchBinder();
    private final StopWatch stopWatch = new StopWatch();
    private NotificationManager mNotificationManager;
    private PowerManager.WakeLock wakeLock;


    public static void checkDeath(Context context) {
        context.startService(new Intent(context, StopWatchService.class));
    }

    public static void start(Context context) {
        context.startService(new Intent(context, StopWatchService.class).putExtra(KEY_ACTION, KEY_START));
    }

    public static void stop(Context context) {
        context.startService(new Intent(context, StopWatchService.class).putExtra(KEY_ACTION, KEY_STOP));
    }


    @Override
    public void onCreate() {
        super.onCreate();
        mNotificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        PowerManager mgr = (PowerManager) getSystemService(Context.POWER_SERVICE);
        wakeLock = mgr.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "MyWakeLock");
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @SuppressLint("WakelockTimeout")
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null && intent.hasExtra(KEY_ACTION)) {
            String action = intent.getStringExtra(KEY_ACTION);
            if (KEY_START.equals(action) && !stopWatch.isStarted()) {
                wakeLock.acquire();
                stopWatch.reset();
                stopWatch.start();
                startForeground(NOTIFICATION_ID, buildNotification(TimeUtils.format(stopWatch.getTime())));
                stopWatch.addOnChronometerTickListener(this);
                return START_STICKY;
            } else if (KEY_STOP.equals(action)) {
                stopWatch.stop();
                stopWatch.removeOnChronometerTickListener(this);
                stopForeground(true);
                if (wakeLock.isHeld()) wakeLock.release();
                return START_NOT_STICKY;
            }
        }
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (wakeLock.isHeld()) wakeLock.release();
    }

    private Notification buildNotification(String time) {
        Intent startMainActivity = new Intent(this, MainActivity.class);
        startMainActivity.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        PendingIntent pi = PendingIntent.getActivity(this, 0, startMainActivity, 0);
        return new NotificationCompat.Builder(this)
                .setContentTitle("Marathon is running: " + time)
                .setContentIntent(pi)
                .setSmallIcon(R.mipmap.ic_launcher)
                .setOngoing(true)
                .build();
    }


    @Override
    public void onStopWatchStarted(long startTime, long nanoStartTime) {}

    @Override
    public void onStopWatchStopped(long stopTime, long nanoStopTime) {}

    @Override
    public void onStopWatchStateChanged(int runningState) {}

    @Override
    public void onStopWatchTick(long time, long nanoTime, long currentNanoTime) {
        mNotificationManager.notify(1337, buildNotification(TimeUtils.format(stopWatch.getTime())));
    }


    class StopWatchBinder extends Binder {
        void addStopWatchListener(StopWatch.OnStopWatchListener listener) {
            stopWatch.addOnChronometerTickListener(listener);
        }

        void removeStopWatchListener(StopWatch.OnStopWatchListener listener) {
            stopWatch.removeOnChronometerTickListener(listener);
        }

        StopWatch getStopWatch() {
            return stopWatch;
        }
    }
}

File: app/src/main/java/ua/hospes/rtm/core/StopWatchService.kt
package ua.hospes.rtm.core

import android.annotation.SuppressLint
import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Binder
import android.os.IBinder
import android.os.PowerManager
import androidx.core.app.NotificationCompat
import ua.hospes.rtm.BuildConfig

import ua.hospes.rtm.R
import ua.hospes.rtm.ui.MainActivity
import ua.hospes.rtm.utils.TimeUtils

private const val NOTIFICATION_ID = 1337
private const val KEY_ACTION = "action"
private const val KEY_START = "start"
private const val KEY_STOP = "stop"

class StopWatchService : Service(), StopWatch.OnStopWatchListener {
    private val binder = StopWatchBinder()
    private val stopWatch = StopWatch()
    private val mNotificationManager: NotificationManager by lazy { getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager }
    private val wakeLock: PowerManager.WakeLock by lazy {
        val mgr = getSystemService(Context.POWER_SERVICE) as PowerManager
        mgr.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "ua.hospes.rtm:MyWakeLock")
    }


    override fun onBind(intent: Intent): IBinder? = binder

    @SuppressLint("WakelockTimeout")
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int = when (intent?.getStringExtra(KEY_ACTION)) {
        KEY_START -> {
            if (!stopWatch.isStarted) {
                wakeLock.acquire()
                stopWatch.reset()
                stopWatch.start()
                startForeground(NOTIFICATION_ID, buildNotification(TimeUtils.format(stopWatch.time)))
                stopWatch.addOnChronometerTickListener(this)
            }
            START_STICKY
        }

        KEY_STOP -> {
            stopWatch.stop()
            stopWatch.removeOnChronometerTickListener(this)
            stopForeground(true)
            if (wakeLock.isHeld) wakeLock.release()
            START_NOT_STICKY
        }

        else -> START_NOT_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        if (wakeLock.isHeld) wakeLock.release()
    }

    private fun buildNotification(time: String): Notification {
        val startMainActivity = Intent(this, MainActivity::class.java)
            .apply { flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP }
        val pi = PendingIntent.getActivity(this, 0, startMainActivity, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        return NotificationCompat.Builder(this, BuildConfig.NOTIFICATION_CHANNEL_RACE)
            .setContentTitle("Marathon is running: $time")
            .setContentIntent(pi)
            .setSmallIcon(R.mipmap.ic_launcher)
            .setOnlyAlertOnce(true)
            .setOngoing(true)
            .build()
    }


    override fun onStopWatchStarted(startTime: Long, nanoStartTime: Long) {}

    override fun onStopWatchStopped(stopTime: Long, nanoStopTime: Long) {}

    override fun onStopWatchStateChanged(runningState: Int) {}

    override fun onStopWatchTick(time: Long, nanoTime: Long, currentNanoTime: Long) {
        mNotificationManager.notify(NOTIFICATION_ID, buildNotification(TimeUtils.format(stopWatch.time)))
    }


    internal inner class StopWatchBinder : Binder() {
        fun addStopWatchListener(listener: StopWatch.OnStopWatchListener) {
            stopWatch.addOnChronometerTickListener(listener)
        }

        fun removeStopWatchListener(listener: StopWatch.OnStopWatchListener) {
            stopWatch.removeOnChronometerTickListener(listener)
        }

        fun getStopWatch(): StopWatch = this@StopWatchService.stopWatch
    }

    companion object {
        @JvmStatic
        fun checkDeath(context: Context) =
            context.startService(Intent(context, StopWatchService::class.java)).let { Unit }

        @JvmStatic
        fun start(context: Context) =
            context.startService(Intent(context, StopWatchService::class.java).putExtra(KEY_ACTION, KEY_START)).let { Unit }

        @JvmStatic
        fun stop(context: Context) =
            context.startService(Intent(context, StopWatchService::class.java).putExtra(KEY_ACTION, KEY_STOP)).let { Unit }
    }
}